### PR TITLE
Addition of levelzero chip energy and power support.

### LINKED ISSUE #2074: LevelZero integration test update for signal rename
https://github.com/geopm/geopm/blob/5c825c2ab04bc39cabb69991afe4cdf25a5b3258/integration/test_skipped/test_levelzero_signals.py#L73

Should be LEVELZERO::GPU_POWER

### PR DESCRIPTION
**Pull request by lhlawson**
- Relates to #1993 bug report from github issues
- Fixes #2074 change request from github issues.

Update to LevelZero Device Pool, IOGroup, and LevelZero.cpp to support chip level energy readings and derivative power signals.  

TODO: 
- [x] refactor energy readings for accelerator and accelerator_chip (currently 2 functions are used when 1 would work).
- [x] Update ordering of GEOPM domain, levelzero domain, idx, etc for all LevelZero.cpp and LevelZeroDevicePool.cpp functions to match (new functions are scattershot in their ordering).
- [x] Add support for any power_limit* functions at the chip level where appropriate.  Pushed to later PR.
- [x] Update power domain tracking for cases where ZES_AFFINITY_MASK is used in a way that forces sub-devices to act as device (i.e. =0.0)

### CODE REVIEW

File: service/src/LevelZero.hpp [hunk 1]
```diff
@@ -158,20 +166,37 @@ namespace geopm
...
             /// @brief Get the LevelZero device energy and timestamp
             ///        in microjoules and microseconds
+            /// @param [in] geopm_domain The GEOPM domain being targeted
             /// @param [in] l0_device_idx The index indicating a particular
             ///        Level Zero accelerator.
+            /// @param [in] l0_domain_idx The index indicating a particular
+            ///        Level Zero domain.
             /// @return Accelerator energy in microjoules and timestamp in microseconds
-            virtual std::pair<uint64_t, uint64_t> energy_pair(unsigned int l0_device_idx) const = 0;
+            /// TODO: document & collapse
```
Comment (bgeltz) [inline service/src/LevelZero.hpp]: What's left for this TODO?

Comment (lhlawson) [reply service/src/LevelZero.hpp]: Removing the TODO is all that's left.

**Comment by lhlawson** [comment]
This PR will run afoul of user ENV settings, specifically the ZE_AFFINITY_MASK setting.  
Assuming 1 device (0) with two chip (0.0, 0.1), the Level Zero API allows the user to control which hardware is exposed.  This will result in GEOPM signals being 'dropped'.  Examples below are from before the GEOPM 2.0 name change (e.g. GPUCHIP_POWER is now GPU_CORE_POWER) 

Default case (ZE_AFFINITY_MASK=0 or ZE_AFFINITY_MASK not set) device enumeration & GEOPM behavior
```
Device 0 is a discrete GPU
        Device 0 has 2 subdevices
        Device 0 has 3 power domains
                 Found 1 power domains with onSubdevice == 0
                 Found 2 power domains with onSubdevice == 1
geopmread LEVELZERO::GPU_POWER board_accelerator 0
<GPU 0 result>
geopmread LEVELZERO::GPUCHIP_POWER board_accelerator_chip 0
<GPU Chip 0 & Chip 1 aggregate result>
geopmread LEVELZERO::GPUCHIP_POWER board_accelerator_chip 0
<GPU Chip 0 result>
geopmread LEVELZERO::GPUCHIP_POWER board_accelerator_chip 1
<GPU Chip 1 result>
```

With a modified affinity mask that forces only the tile to be used:
```
export ZE_AFFINITY_MASK=0.0
Device 0 is a discrete GPU
        Device 0 has 0 subdevices
        Device 0 has 1 power domains
                 Found 1 power domains with onSubdevice == 0
                 Found 0 power domains with onSubdevice == 1

geopmread LEVELZERO::GPUCHIP_POWER board_accelerator 0
Error: cannot read signal: <geopm> Invalid argument: PlatformIOImp::read_signal(): no support for signal name "LEVELZERO::GPUCHIP_POWER" and domain type "10"
The following errors were observed:
<geopm> Invalid argument: LevelZeroDevicePool::energy: Not supported on this hardware for the specified LevelZero domain
...

geopmread LEVELZERO::GPUCHIP_POWER board_accelerator_chip 0
Error: cannot read signal: <geopm> Invalid argument: PlatformIOImp::read_signal(): no support for signal name "LEVELZERO::GPUCHIP_POWER" and domain type "10"
The following errors were observed:
<geopm> Invalid argument: LevelZeroDevicePool::energy: Not supported on this hardware for the specified LevelZero domain
...

geopmread LEVELZERO::GPU_POWER board_accelerator 0
<GPU  0 result>
```
Note that in this mode we still provide the entire GPU the GPU_POWER.  This appears to match the Level Zero API enumeration change on the hardware.  

Edited final behavior after testing on a wider variety of hardware.

File: integration/test_skipped/test_levelzero_signals.py [hunk 2]
```diff
@@ -5,6 +5,8 @@
 #
 
 
+from __future__ import absolute_import
+
```
Comment (bgeltz) [inline integration/test_skipped/test_levelzero_signals.py]: Why is this necessary?

File: service/src/LevelZero.cpp [hunk 3]
```diff
@@ -248,22 +249,33 @@ namespace geopm
...
-#endif
+            }
 
-                if (num_device_power_domain != 1) {
-                    throw Exception("LevelZero::" + std::string(__func__) +
-                                    ": GEOPM requires a single device level power domain, "
-                                    "but found " + std::to_string(num_device_power_domain),
-                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
-                }
+            if (num_device_power_domain != 1 && num_subdevice_power_domain == 0) {
```
Comment (bgeltz) [inline service/src/LevelZero.cpp]: If a system has multiple device power domains and at least one subdevice power domain somewhere, this will not throw.  Could a system be configured that way?  If yes, this needs to be more robust.

File: service/src/LevelZero.cpp [hunk 4]
```diff
@@ -508,48 +536,99 @@ namespace geopm
...
-        return m_devices.at(l0_device_idx).cached_energy_timestamp;
+        uint64_t timestamp = 0;
+        //PACKAGE
+        if(geopm_domain == GEOPM_DOMAIN_GPU) {
+            timestamp = m_devices.at(l0_device_idx).cached_energy_timestamp;
+        }
+        else if(geopm_domain == GEOPM_DOMAIN_GPU_CHIP) {
+        //TILE
+            timestamp = m_devices.at(l0_device_idx).subdevice.cached_energy_timestamp.at(l0_domain_idx);
+        }
```
Comment (bgeltz) [inline service/src/LevelZero.cpp]:
```suggestion
        //PACKAGE
        if(geopm_domain == GEOPM_DOMAIN_GPU) {
            timestamp = m_devices.at(l0_device_idx).cached_energy_timestamp;
        }
        //TILE
        else if(geopm_domain == GEOPM_DOMAIN_GPU_CHIP) {
            timestamp = m_devices.at(l0_device_idx).subdevice.cached_energy_timestamp.at(l0_domain_idx);
        }
```

File: service/src/LevelZero.cpp [hunk 5]
```diff
@@ -508,48 +536,99 @@ namespace geopm
...
+        //TODO: either check l0_domain, or move to treating it the same
+        //      as all other signals (i.e. All, Compute, and Mem)
```
Comment (bgeltz) [inline service/src/LevelZero.cpp]: Is this implementation complete?

**Review by bgeltz — CHANGES_REQUESTED** [review]
Would you also please rebase this on dev?

File: service/src/LevelZero.cpp (hunk 3 above)
Comment (lhlawson) [reply service/src/LevelZero.cpp]: There is a prior check ~line 252 for:
 if (num_device_power_domain != 1) {          
That should prevent multiple device level power domains from being tracked.  We could consolidate these various checks if needed.

File: service/src/LevelZero.cpp (hunk 5 above)
Comment (lhlawson) [reply service/src/LevelZero.cpp]: These checks have been moved to the LevelZeroDevicePool.  Latest commit removes the TODOs.

File: service/src/LevelZero.cpp (hunk 3 above)
Comment (bgeltz) [reply service/src/LevelZero.cpp]: Consider if the ```power_domain``` vector had the following entries:
```
    power_domain[0] = Has subdevice power domain
    power_domain[1] = Has device power domain
```

On the first pass through the for loop on line ~240, the ```else()``` case is hit and ```num_subdevice_power_domain``` is incremented.  On the next pass, the ```if()``` case is hit, ```num_device_power_domain``` is incremented, and the throw on line ~252 is skipped.  When we get here to line ~265, we have 1 ```device_power_domain``` and 1 ```subdevice_power_domain```.  Is this an error condition?

If you want to sync on this and try to iron out all these error conditions and consolidate the checks, let me know.

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: integration/test_skipped/test_levelzero_signals.py (hunk 2 above)
Comment (bgeltz) [inline integration/test_skipped/test_levelzero_signals.py]:
```suggestion
```
Needed?

File: service/src/LevelZeroDevicePool.cpp [hunk 6]
```diff
@@ -290,41 +290,98 @@ namespace geopm
...
+            energy_timestamp = m_levelzero.energy_timestamp(domain, domain_idx,
+                                                            l0_domain, 0);
+        }
+        else if (domain == GEOPM_DOMAIN_GPU_CHIP) {
+            std::pair<unsigned int, unsigned int> dev_subdev_idx_pair;
+            dev_subdev_idx_pair = subdevice_device_conversion(domain_idx);
+
+            check_domain_exists(m_levelzero.power_domain_count(domain,
+                                            dev_subdev_idx_pair.first, l0_domain),
+                                            __func__, __LINE__);
```
Comment (bgeltz) [inline service/src/LevelZeroDevicePool.cpp]: Indentation

File: service/src/LevelZeroDevicePool.cpp [hunk 7]
```diff
@@ -290,41 +290,98 @@ namespace geopm
...
                                                       unsigned int domain_idx,
                                                       int l0_domain) const
     {
-        if (domain != GEOPM_DOMAIN_GPU) {
+        uint64_t energy_timestamp = 0;
+        if (domain == GEOPM_DOMAIN_GPU) {
+            check_idx_range(domain, domain_idx);
+            check_domain_exists(m_levelzero.power_domain_count(domain,
+                                            domain_idx, l0_domain),
+                                            __func__, __LINE__);
```
Comment (bgeltz) [inline service/src/LevelZeroDevicePool.cpp]: Indentation issue

File: service/src/LevelZeroDevicePool.cpp [hunk 8]
```diff
@@ -290,41 +290,98 @@ namespace geopm
...
                                                                       int l0_domain) const
     {
-        if (domain != GEOPM_DOMAIN_GPU) {
+        std::pair<uint64_t, uint64_t> result = {0,0};
+
+        if (domain == GEOPM_DOMAIN_GPU) {
+            check_idx_range(domain, domain_idx);
+            check_domain_exists(m_levelzero.power_domain_count(domain,
+                                            domain_idx, l0_domain),
+                                            __func__, __LINE__);
```
Comment (bgeltz) [inline service/src/LevelZeroDevicePool.cpp]: Indentation issue

File: service/src/LevelZeroDevicePool.cpp [hunk 9]
```diff
@@ -290,41 +290,98 @@ namespace geopm
...
+
+            result = m_levelzero.energy_pair(domain, domain_idx, -1);
+        }
+        else if (domain == GEOPM_DOMAIN_GPU_CHIP){
+            std::pair<unsigned int, unsigned int> dev_subdev_idx_pair;
+            dev_subdev_idx_pair = subdevice_device_conversion(domain_idx);
+            check_domain_exists(m_levelzero.power_domain_count(domain,
+                                            dev_subdev_idx_pair.first, l0_domain),
+                                            __func__, __LINE__);
```
Comment (bgeltz) [inline service/src/LevelZeroDevicePool.cpp]: Indentation issue

File: service/src/LevelZeroDevicePool.cpp [hunk 10]
```diff
@@ -290,41 +290,98 @@ namespace geopm
...
+            check_idx_range(domain, domain_idx);
+            energy = m_levelzero.energy(domain, domain_idx, l0_domain, 0);
+        }
+        else if (domain == GEOPM_DOMAIN_GPU_CHIP) {
+            std::pair<unsigned int, unsigned int> dev_subdev_idx_pair;
+            dev_subdev_idx_pair = subdevice_device_conversion(domain_idx);
+
+            check_domain_exists(m_levelzero.power_domain_count(domain,
+                                            dev_subdev_idx_pair.first, l0_domain),
+                                            __func__, __LINE__);
```
Comment (bgeltz) [inline service/src/LevelZeroDevicePool.cpp]: Indentation issue

File: service/src/LevelZero.cpp [hunk 11]
```diff
@@ -509,48 +543,95 @@ namespace geopm
...
-        return energy_pair(l0_device_idx).first;
+        return energy_pair(geopm_domain, l0_device_idx, l0_domain_idx).first;
     }
 
-    std::pair<uint64_t,uint64_t> LevelZeroImp::energy_pair(unsigned int l0_device_idx) const
+    std::pair<uint64_t,uint64_t> LevelZeroImp::energy_pair(int geopm_domain,
+                                                           unsigned int l0_device_idx,
+                                                           int l0_domain_idx) const
     {
         ze_result_t ze_result;
```
Comment (bgeltz) [inline service/src/LevelZero.cpp:573]: I'm wondering now if we should just initialize all structs with {} that are ever pass-by-reference in a new PR...

**Review by bgeltz — CHANGES_REQUESTED** [review]

**Review by bgeltz — DISMISSED** [review]

**Comment by bgeltz** [comment]
> This PR will run afoul of user ENV settings, specifically the ZE_AFFINITY_MASK setting...

Should we include some docu about this?

**Comment by lhlawson** [comment]
> > This PR will run afoul of user ENV settings, specifically the ZE_AFFINITY_MASK setting...
> 
> Should we include some docu about this?

The final observed behavior appeared to be that the power domain would be consistently exposed to the user, but pointing users to https://spec.oneapi.io/level-zero/latest/core/PROG.html#environment-variables isn't a bad idea.

Added in latest commit.

File: service/docs/source/geopm_pio_levelzero.7.rst [hunk 12]
```diff
@@ -12,6 +12,8 @@ Requirements
 
 To use the GEOPM LevelZero signals and controls GEOPM must be compiled against the oneAPI LevelZero libraries and must be run on a system with discrete GPUs supported by LevelZero.  To compile against the oneAPI LevelZero libraries geopm must be configured using the --enable-levelzero flag.  The optional --with-levelzero flag may be used to indicate the path of the required libraries.  In addition the user must export ZES_ENABLE_SYSMAN=1 as specified by the Intel oneAPI Level Zero Sysman documentation.  See the Sysman specification for more info on related environment variables and their usage.
 
+Since signals and controls are exposed via the Sysman API they will be impacted by Sysman environment variables.  Please review `oneAPI LevelZero Sysman Environment Variables <https://spec.oneapi.io/level-zero/latest/sysman/PROG.html#environment-variables>`_ and `oneAPI LevelZero Core Programming Guide Environment Variable <https://spec.oneapi.io/level-zero/latest/core/PROG.html#environment-variables>`_.
```
Comment (bgeltz) [inline service/docs/source/geopm_pio_levelzero.7.rst]:
```suggestion
Since signals and controls are exposed via the Sysman API they will be impacted by Sysman environment variables.  Please review `oneAPI LevelZero Sysman Environment Variables <https://spec.oneapi.io/level-zero/latest/sysman/PROG.html#environment-variables>`_ and `oneAPI LevelZero Core Programming Guide Environment Variables <https://spec.oneapi.io/level-zero/latest/core/PROG.html#environment-variables>`_.
```

Missing s.

**Review by bgeltz — CHANGES_REQUESTED** [review]
Looks good otherwise.

**Review by bgeltz — APPROVED** [review]
